### DISCUSSION TITLE
关于ecs的一些疑问

### DISCUSSION THREAD

**Discussion by constly** [ejoy/ant#70] [General]
1. 假如我在package.ecs中定义了一个system
```
system "main_system_a" .implement "main_system.lua"
```
然后在main_system.lua中
```
local ecs = ...
local m = ecs.system "main_system_b"
```
我发现在.ecs中预定义的system名字与.lua中实际创建的名字是没啥关联的，实际上在.lua中还可以创建多个system；
另外，world:enable_system() 与 world:disable_system() 的参数都与预定义的名字无关；
请问这怎么理解呢？.ecs中的预定义是否可以随机生成，直接简化为: system_implement "main_system.lua"。

2. 我看到很多system的写法都是
```
function m:data_changed()
end
```
这让我以为函数内部有self可以用，但实际用self会出问题，比如你通过self调用文件内另外一个函数会报错。总之，觉得有点迷惑。
是否应该尽量统一写法为:
```
function m.data_changed()
end
```

**Comment by cloudwu** [comment]
你需要了解的是 pipeline 和 stage 概念。 https://github.com/ejoy/ant/wiki/ECS

上面提到的 data_changed 是一个 stage 的名字。

system 的名字在运行时是没有什么意义的，stage 才有意义。

stage 函数中的 self 指的是 world 对象。对 C 实现的 system 是有意义的。

**Reply by cloudwu** [reply]
.ecs 是一个描述文件，在框架做依赖分析的时候有用。所以需要在里面写明所有信息。如果把 system 名字只写在实现中，就无法只从 .ecs 中得到所有元信息了。

**Reply by constly** [reply]
我觉得system的名字在运行时是有意义的，我刚才都还在找怎么通过system对象得到他的名字，
最后没找到，只能把这句代码"local m = ecs.system 'main_system'" 封装一遍。

意义主要在于，目前world的enable_system和disable_system要用到这个名字，这个两个接口在游戏开发当中，有很大概率会用到 = =

**Reply by constly** [reply]
wiki中ecs那篇，我看了两三遍了，只是感觉离完全理解，还得多实践下 = =

**Comment by actboy168** [comment]
1. 这是一个历史遗留问题，ecs文件历经了多次的重构和简化。也许应该把ecs文件中的system去掉，也许应该强校验ecs定义的system和lua定义的system，这个没有定论。保留现在的样子是一个较好的折中方案。

2. 有self就说明有它存在的作用，只是这个self不是你想象中的那个值而已。当然你不需要它时，写m:data_changed还是写m.data_changed都是无妨的。

**Reply by constly** [reply]
好的，了解~
老实说，这个self指的是world对象，实在是有点隐晦 = =

**Reply by actboy168** [reply]
隐晦的是使用的人，而不是api
